File: SudokuSolver.py
def main():
    printDummy()
    print("Enter the numbers of Sudoku as shown above:")
    grid = inputGrid()
    print("\nUnsolved Sudoku:")
    printGrid(grid)
    if solve(grid):
        print("\nSolved Sudoku:")
        printGrid(grid)
    else:
        print("\nThis Sudoku can't be solved")


def printGrid(grid):
    for row, Row in enumerate(grid):
        for col, num in enumerate(Row):
            print(num, end=" ")
            if col % 3 == 2 and col != 8:
                print("|", end=" ")
        print("")
        if row % 3 == 2 and row != 8:
            print("---------------------")


def printDummy():
    for i in range(9):
        for j in range(9):
            print(0, end=" ")
        print("")


def inputGrid():
    grid = [input().split(" ") for _ in range(9)]
    grid = [[int(num) for num in Row] for Row in grid]
    return grid


def findEmpty(grid):
    for row, Row in enumerate(grid):
        for col, num in enumerate(Row):
            if num == 0:
                return row, col
    return None


def isValid(grid, num, pos):
    row, col = pos
    for i in range(9):
        if grid[row][i] == num:
            return False
        if grid[i][col] == num:
            return False

    ROW = row // 3
    COL = col // 3
    for i in range(3):
        for j in range(3):
            if grid[3 * ROW + i][3 * COL + j] == num:
                return False
    return True


def solve(grid):
    pos = findEmpty(grid)
    if not pos:
        return True
    row, col = pos

    for num in range(1, 10):
        if isValid(grid, num, pos):
            grid[row][col] = num
            if solve(grid):
                return True
            grid[row][col] = 0
    return False


main()

File: SudokuSolver.java
import java.util.Scanner;

public class SudokuSolver {
    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        printDummy();
        System.out.println("Enter the numbers of Sudoku as shown above:");
        int[][] grid = inputGrid(scanner);
        System.out.println("\nUnsolved Sudoku:");
        printGrid(grid);
        if (solve(grid)){
            System.out.println("\nSolved Sudoku:");
            printGrid(grid);
        }
        else
            System.out.println("\nThis Sudoku can't be solved");
    }


    public static void printGrid(int[][] grid){
        for (int row = 0; row <= 8; row++){
            for (int col = 0; col <= 8; col++){
                System.out.printf("%d ", grid[row][col]);
                if (col % 3 == 2 && col != 8)
                    System.out.print("| ");
            }
            System.out.print("\n");
            if (row % 3 == 2 && row != 8)
                System.out.println("---------------------");
        }
    }


    public static void printDummy(){
        for (int i = 0; i <= 8; i++){
            for (int j = 0; j <= 8; j++){
                System.out.printf("%d ", 0);
            }
            System.out.print("\n");
        }
    }


    public static int[][] inputGrid(Scanner scanner){
        String[][] rawGrid = new String[9][9];
        for (int row = 0; row <= 8; row++){
            rawGrid[row] = scanner.nextLine().split(" ", 9);
        }

        int[][] grid = new int[9][9];
        for (int row = 0; row <= 8; row++){
            for (int col = 0; col <= 8; col++){
                grid[row][col] = Integer.parseInt(rawGrid[row][col]);
            }
        }
        return grid;
    }


    public static int[] findEmpty(int[][] grid){
        int[] pos = {9, 9};
        for (int row = 0; row <= 8; row++){
            for (int col = 0; col <= 8; col++){
                if (grid[row][col] == 0){
                    pos[0] = row;
                    pos[1] = col;
                    return pos;
                }
            }
        }
        return pos;
    }


    public static boolean isValid(int[][] grid, int num, int[] pos){
        int row = pos[0];
        int col = pos[1];
        for (int i = 0; i <= 8; i++){
            if (grid[row][i] == num)
                return false;
            if (grid[i][col] == num)
                return false;
        }

        int ROW = row / 3;
        int COL = col / 3;
        for (int i = 0; i<= 2; i++){
            for (int j = 0; j <= 2; j++){
                if (grid[3 * ROW + i][3 * COL + j] == num)
                    return false;
            }
        }
        return true;
    }


    public static boolean solve(int[][] grid){
        int[] pos = findEmpty(grid);
        int row = pos[0];
        int col = pos[1];
        if (row == 9 && col == 9)
            return true;

        for (int num = 1; num <= 9; num++){
            if (isValid(grid, num, pos)){
                grid[row][col] = num;
                if (solve(grid))
                    return true;
                grid[row][col] = 0;
            }
        }
        return false;
    }
}
